sdk/java/pulumi/src/test/java/com/pulumi/deployment/DeploymentRunnerTest.java

```diff
@@ -46,19 +47,26 @@ void testTerminatesEarlyOnException() {
         assertThat(result.resources).isNotNull();
         assertThat(result.resources).isNotEmpty();
         assertThat(result.resources).hasSize(1);
-        var stack = (TerminatesEarlyOnExceptionStack) result.resources.get(0);
+        var stack = TerminatesEarlyOnExceptionStack.of(result);
         assertThat(Internal.of(stack.slowOutput).getDataAsync()).isNotCompleted();
         assertThat(Internal.of(stack.slowOutput).getValueNullable()).isNotCompleted();
     }
 
-    public static class TerminatesEarlyOnExceptionStack extends Stack {
-        @Export(type = Integer.class)
+    public static class TerminatesEarlyOnExceptionStack {
         public final Output<Integer> slowOutput;
 
-        public TerminatesEarlyOnExceptionStack() {
+        public TerminatesEarlyOnExceptionStack(Output<Integer> slowOutput) {
+            this.slowOutput = slowOutput;
+        }
+
+        public static void init(Context ctx) {
             Output.of(CompletableFuture.failedFuture(new RunException("Deliberate test error")));
-            this.slowOutput = Output.of(new CompletableFuture<Integer>()
-                    .completeOnTimeout(1, 60, TimeUnit.SECONDS));
+            ctx.export("slowOutput",Output.of(new CompletableFuture<Integer>()
+                    .completeOnTimeout(1, 60, TimeUnit.SECONDS)));
+        }
+
+        public static TerminatesEarlyOnExceptionStack of(DeploymentTests.DeploymentMock.TestAsyncResult result) {
+            return new TerminatesEarlyOnExceptionStack(result.getStackOutput("slowOutput"));
```

Comment:
NOTE: we used to have reflection via `@Export` annotation help consolidate that the types and slots of "export" declarations align with the fields in the Stack object, which is helpful in the actual test code. I wanted to keep that typing benefit here but rolled boilerplate hand-written code to recover the types in the `.of()` methods. Not particularly pretty but gets this done.

Reply:
I'd much rather we add good code, instead of even worse code then we're trying to refactor.
Getting those values from Stack instance would be a lot better IMO.

Reply:
How would that work?

We're inching toward a world where Stack is a final class and is only used in internally.

The only alternative here is a reflecting facility that hydrates a class out of `Map<String, Output<T>>` which we do not have in the codebase I think. I don't think adding such a facility would belong in this PR,  and if the only motivation is so that we keep the test suite, that is a weak motivation.

I think a good place to return to this is when re-designing the testing API for the users. This code here is fine for what it's doing.

Reply:
I might have a simple idea, but I need to test it first.